Add `Mithril Threat Model` page on website (#1726)
input-output-hk/mithril: docs/website/root/mithril/threat-model.md

The aggregator repeatedly checks whether enough valid signatures (to reach the quorum) are available to aggregate a Mithril stake-based multi-signature into a certificate.

Mithril certificates are certifying some chain data using an aggregated multi-signature verification key and are chained up to some genesis certificate, which is signed by a genesis signing key.

Mithril clients do connect to an aggregator using HTTP over TLS to query mithril certificates for certified chain data and locate artifacts.

A mithril client can verify the received mithril certificate is linked to other certificates up to the genesis certificate and can be verified using the Mithril genesis verification key (see [details](https://mithril.network/doc/mithril/mithril-protocol/certificates/)).

:::info To do

Missing: the currently recommended relay (reverse proxy)

:::

#### Deployment architecture

This document is specifically targeting the standard deployment architecture where a mithril signer runs next to the block producing node, while access to the mithril aggregator is only done through a relay.

> [![Mithril - Container View](./images/deployment-architecture.jpg)](./images/deployment-architecture.jpg)

Author: The schema is referring to the experimental P2P relay implementation which is not deployed to production.
Maybe a link to these docs is enough:
-  [architecture](https://mithril.network/doc/mithril/mithril-network/architecture#architecture-diagram) 
- [deployment model](https://mithril.network/doc/manual/getting-started/run-signer-node#mithril-signer-deployment-model) in the SPO guide

WDYT?

### External Dependencies

Listing dependencies for mithril-signer
```
% cargo-deps-list -e normal -p mithril-signer 
...
slog-json v2.6.1 {default}
slog-scope v4.4.0
arc-swap v1.6.0

Total dependencies: 267
```

<!-- * [libp2p](https://docs.libp2p.io/)
    * https://docs.libp2p.io/concepts/security/security-considerations/

>     * [DoS Mitigation](https://docs.libp2p.io/concepts/security/dos-mitigation/) -->

Author: We should probably remove all mentions to `libp2p` which is used only for experimentation and not in production?

* Forge individual signatures impersonating one of the signers

## Assets

For each asset we first identify what protection is required: Availability, Confidentiality, Integrity ie. the [CIA Triad](https://www.splunk.com/en_us/blog/learn/cia-triad-confidentiality-integrity-availability.html)

:::info Note

I follow Pascal's suggestion to identify assets as processes rather than particular resources, except in the case of actual pieces of data (eg. keys). The latter should be listed in the threats section and their protection mitigated

:::

We then identify threats and countermeasures

#### KES private keys

* KES key is present only on BP Node but needs to be shared with both the cardano-node process and the mithril-signer process
* KES keys are needed by mithril-signer in order to sign verification key along with an operational certificate which authenticates the key from this stake pool id
* This signing happens at every epoch
    * **availability**: Yes (If KES key is unavailable then signing cannot proceeed)
    * **confidentiality**: Yes (Capturing KES private keys allow an attacker to impersonate a registered SPO on-chain and produce blocks on his behalf)

>     * **integrity**: Yes (Rotating a compromised KES key is a time-consuming process, not sure if it's even possible if the KES period has not passed completely? See https://github.com/input-output-hk/cardano-node-wiki/blob/main/docs/stake-pool-operations/7_KES_period.md)

Author: Yes, KES keys can be rotated when needed, even if the KES period has not expired yet.

Reviewer: We should write this sentence into the threat model then.

Author: I modified the sentence with the fact that KES keys even if they have not expired :+1:

#### Cardano Chain Database

A cardano-node maintains an on-disk database consisting of the chain's history. This database is updated by the node when new blocks are diffused through the network, or minted, and also contains a cache of the ledger state. 

Mithril signer needs needs access to _trusted_ and _up-to-date_ Chain database in order to be able to sign snapshots. 

* **availability**: Yes
* **integrity**: Yes
* **Confidentiality**: No (Data is public and replicated) 

#### Cardano Ledger state

Access to an accurate ledger state is needed by mithril signer to retrieve reliable _Stake distribution_. This access is currently done through a local connection (direct w/ Pallas or indirect with cardano-cli) to a trusted cardano-node

* **availability**: Yes (without SD, signer cannot register keys nor validly use other signers' keys)
* **integrity**: Yes (same, inaccurate SD will make key registration and signing process invalid)
* **confidentiality**: No

#### Mithril signing keys

SPOs register their Mithril keys every epoch to be able to sign snapshots. An attacker could impersonate the SPO and sign invalid snapshots if they got hold of those keys

> Signing keys are currently stored on-disk (?) 

Author: Yes, as they are used `2` epochs after their creation, and deleted `2` epochs after they have been used.
Storage is not encrypted currently. (Should probably be?)

Reviewer: Is this an answer to the `(?)`? If yes, please add this information in

Author: Done

* **confidentiality**: Yes 
* **integrity**: Yes (invalid key is useless obviously)
* **availability**: Yes (Signer needs Key at every signing round, unavailability will lead to inability to sign)

#### Mithril signing keys registration

Mithril signer needs to register new key every epoch with aggregator (and ultimately other signers)

> * **confidentiality**: Yes (access to a signer's key will allow an attacker to impersonate a signer for the duration of the epoch)

Author: ```suggestion
* **confidentiality**: No (only verification keys and proofs of possession are used in the process) 
```

Reviewer 2: The registration takes vk and pop; however, vk is generated by the signing key. In this case, we may make clear that the confidentiality of vk and pop is not necessary, but the signing key is.

Reviewer: We should keep the `(access to a signer's key will allow an attacker to impersonate a signer for the duration of the epoch)` as an explanation why confidentiality of the signing key itself is important (one section above).

Author: IMO we should probably make a clear separation between the two sections:
- Mithril signing keys (private information, must be kept secret)
- Mithril verification keys (public information, must be registered with the other signers)
WDYT?

Author: Done

#### Mithril protocol parameters 

Protocol parameters are needed to coordinate the production of valid multi-signatures. They are served by the aggregator

* **confidentiality**: No (they actually need to be public)
* **availability**: Yes (partial?)
* **integrity**: Yes (tampering protocol parameters can lead a signer to produce invalid signatures)

#### Era reader configuration files

The era reader [Era address](https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/release-mainnet/era.addr) and [Era vkey](https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/release-mainnet/era.vkey) are used by signers to extract information about the current Mithril Era which defines the structure of snapshots and therefore signatures

#### Era reader signing key

The key tied to the aforementioned era address 

### Client-side only assets

These are the assets that are relevant only when downloading and verifying certificates and full-node snapshots. 

:::info To do

Verification process being trustless, does it really make sense to include those in the Threat Model?

> :::

Author: This could be removed as the verification is trustless :+1:

Author: Done

#### Mithril certificate verification process

Mithril clients download snapshots and verify associated certificates using the mithril-client library, either from a CLI tool or [embedded in a browser](https://mithril.network/explorer/) 

* **availability**: Yes (?)
* **confidentiality**: No
* **integrity**: No (the snapshots and certificates are assumed to be secure, integrity is inherent to the protocol)

#### Mithril snapshots & certificates

Mithril certificates are produced by the aggregator from the individual signatures. Mithril certificates' security rests upon a chain of trust ultimately pointing at the Genesis certificate. Mithril aggregator maintains snapshots and certificates (aggregated multi-signatures) to be served to clients

* **availability**: No (Snapshots and certificates are just a fallback, a cardano-node client can always retrieve the infromation from the cardano network itself, albeit much more slowly)
* **confidentiality**: No

> * **integrity**: No (the snapshots and certificates are assumed to be secure, integrity is inherent to the protocol)

Author: Maybe this could be limited to certificates only?

Reviewer 3: A separate section for the snapshot (or more general the artifacts ?) could make sense too.

Author: I added a new section for "Mithril artifacts" :+1:

:::

System consists of three main components: signers, aggregator and clients.

Mithril signers do certify Cardano chain data using a Mithril signing key. That means, Mithril signers need access to a trusted `cardano-node` and the Mithril signing key to operate.

Mithril signing keys are rotated every epoch and need to be certified by the Cardano KES key. For this, the Mithril signers need access to the KES key in order to register the signing key for this epoch. 

Cardano KES keys are also used by block producing `cardano-node` and are typically located on the block producing machine.

Cardano KES keys need to be [evolved every 36 hours](https://github.com/input-output-hk/cardano-node-wiki/blob/main/docs/stake-pool-operations/7_KES_period.md), while they can be rotated from a root key when needed.

:::info To do

Is there a Cardano threat model about this?

:::

All Mithril signers and Mithril clients connect to a single aggregator using HTTP over TLS.

Registering a Mithril signing key means that a signer sends its corresponding verification key to the aggregator, for the purpose of distribution to all other Mithril signers.

> A Mithril aggregator coordinates creation of signatures by all registered signers. Mithril signers do ask the aggregater whether a signature is pending on a regular basis. The aggregator responds with information what to sign and a list of public information about all registered signers.

Reviewer 4: typo: ... Mithril signers do ask the **aggregator** whether...

Author: Done

Each Mithril signer verifies the information, produces a signature of the requested information to sign and submits that to the aggregator (which verifies the signature being correct upon receiving).

The aggregator repeatedly checks whether enough valid signatures (to reach the quorum) are available to aggregate a Mithril stake-based multi-signature into a certificate.

Mithril certificates are certifying some chain data using an aggregated multi-signature verification key and are chained up to some genesis certificate, which is signed by a genesis signing key.

Mithril clients do connect to an aggregator using HTTP over TLS to query Mithril certificates for certified chain data and locate artifacts.

A Mithril client can verify the received Mithril certificate is linked to other certificates up to the genesis certificate and can be verified using the Mithril genesis verification key (see [details](https://mithril.network/doc/mithril/mithril-protocol/certificates/)).

:::info To do

Missing: the currently recommended relay (reverse proxy)

:::

#### Deployment architecture

> This document is specifically targeting the standard deployment architecture where a Mithril signer runs next to the block producing node, while access to the Mithril aggregator is only done through a relay.

Reviewer 2: can include a `block producing (BP) node`, useful for later usages.

Author: Could you elaborate this comment @curiecrypt?

[![Mithril - Container View](./images/deployment-architecture.jpg)](./images/deployment-architecture.jpg)

### External Dependencies

Listing dependencies for mithril-signer
```
% cargo-deps-list -e normal -p mithril-signer 
...
slog-json v2.6.1 {default}
slog-scope v4.4.0
arc-swap v1.6.0

Total dependencies: 267
```

<!-- * [libp2p](https://docs.libp2p.io/)
    * https://docs.libp2p.io/concepts/security/security-considerations/
    * [DoS Mitigation](https://docs.libp2p.io/concepts/security/dos-mitigation/) -->
    
* SPOs infrastructure:
    * Block producing host configuration
    * Relay hosts configuration
    * Firewall/private network

### Entry Points 

> * mithril-relay P2P TCP ports

Author: There is no P2P in production, I removed it.

* mithril-aggregator HTTP port
* mithril-relay HTTP port

## Assumptions 

> The core Mithril protocol is considered safe and its analysis is out of scope for this document, it is computationally infeasible to:

Reviewer 2: We can put the link to the paper

Author: Done

* Forge a valid aggregate signature from forged signing keys
* Forge individual signatures impersonating one of the signers

## Assets

For each asset we first identify what protection is required: Availability, Confidentiality, Integrity ie. the [CIA Triad](https://www.splunk.com/en_us/blog/learn/cia-triad-confidentiality-integrity-availability.html)

:::info Note

> I follow Pascal's suggestion to identify assets as processes rather than particular resources, except in the case of actual pieces of data (eg. keys). The latter should be listed in the threats section and their protection mitigated

Reviewer 5: Is this `info` section still necessary?

Author: It can be removed IMO :+1:

Author: Done

> :::

Author: Done

We then identify threats and countermeasures

#### KES private keys

* KES key is present only on BP Node but needs to be shared with both the cardano-node process and the mithril-signer process
* KES keys are needed by mithril-signer in order to sign verification key along with an operational certificate which authenticates the key from this stake pool id
* This signing happens at every epoch
    * **availability**: Yes (If KES key is unavailable then signing cannot proceeed)

>     * **confidentiality**: Yes (Capturing KES private keys allow an attacker to impersonate a registered SPO on-chain and produce blocks on his behalf)

Reviewer: Should we distinguish between the current KES key from the cold keys? AFAIK does breach on confidentiality of the current KES key allow to impersonate an SPO **until rotated**. That would limit the risk temporally

Author: As we only use KES keys and Operational Certificate, I guess that it is not needed to mention the Cold keys. 
However, we should probably add a sub section for the Operational Certificate, WDYT?

Also, you're right as the impersonation of an SPO is possible only until the KES keys are rotated, I updated the sentence in that sense :+1:

    * **integrity**: Yes (Rotating a compromised KES key is a time-consuming process, not sure if it's even possible if the KES period has not passed completely? See https://github.com/input-output-hk/cardano-node-wiki/blob/main/docs/stake-pool-operations/7_KES_period.md)

#### Block diffusion

Block diffusion process ensures the timely diffusion of blocks, both those produced "locally" and those received from upstream peers
* **availability**: Yes (Not being to diffuse blocks harms a BP's SPO economic viability, and can also harm their ability to create new blocks)
* **confidentiality**: No (block diffusion happens in the open)
* **integrity**: Yes (partially? not sure what integrity means here)

#### Block production

Block production is the process of "minting" new blocks by block producers, driven by Stake-based random lottery.

> The mithril-signer necessarily runs on the same host than a BP because it needs access to the KES signing key

Reviewer 6: ```suggestion
The mithril-signer necessarily runs on the same host as a BP because it needs access to the KES signing key
```

> The mithril-signer necessarily runs on the same host than a BP because it needs access to the KES signing key

Reviewer 6: ```suggestion
The mithril-signer necessarily runs on the same host as a BP because it needs access to the KES signing key
```

Author: Done

* **availability**: Yes (BP is critical for SPOs revenue, and preventing a BP from producing blocks can harm SPOs capabilities to operate)
* **integrity**: Yes (incorrect or invalid data can hamper BP capabilities)
* **confidentiality**: Yes? (BP schedule is private information for the BP, leaking it could provide adversaries advance knowledge of schedules and lead to **grinding attacks** to attempt to manipulate nonce in the disfavour of a BP ??)

#### Cardano Chain Database

A cardano-node maintains an on-disk database consisting of the chain's history. This database is updated by the node when new blocks are diffused through the network, or minted, and also contains a cache of the ledger state. 

> Mithril signer needs needs access to _trusted_ and _up-to-date_ Chain database in order to be able to sign snapshots. 

Reviewer 4: typo: Mithril signer **needs needs** access

Author: Done

> Mithril signer needs needs access to _trusted_ and _up-to-date_ Chain database in order to be able to sign snapshots. 

Reviewer: ```suggestion
Mithril signer needs access to _trusted_ and _up-to-date_ Chain database in order to be able to sign snapshots. 
```

Author: Done

> * **availability**: Yes

Reviewer: ```suggestion
* **availability**: Yes (mithril-signer could corrupt the chain db resulting in down-time)
```

Author: I think you are right, but this means that we should add a new threat in the "Threat & Mitigations" section which would be "Integrity of the Cardano block producer database". WDYT?

> * **integrity**: Yes

Reviewer: How? I'm not entirely clear how the chain db is checked for integrity on each start, but potentially corrupted integrity could make the node have wrong ledger state / stake distributions and consequently impact block production?

Author: Yes, Cardano node database corruption would lead to down-time and/or impact the block production :+1:

Author: I added a new threat section regarding integrity of the block producer database

> * **Confidentiality**: No (Data is public and replicated) 

Reviewer 6: ```suggestion
* **confidentiality**: No (Data is public and replicated) 
```

Author: Done

#### Cardano Ledger state

> Access to an accurate ledger state is needed by Mithril signer to retrieve reliable _Stake distribution_. This access is currently done through a local connection (direct w/ Pallas or indirect with cardano-cli) to a trusted cardano-node

Reviewer: The ledger state / stake distribution is also used by the cardano-node to determine leader schedules and hence corruption here has an impact on the block production process.

Author: Access by the signer is done through local state query mini-protocols of the Cardano node (via local connection or Cardano cli). IMO, there is no impact on the leader schedule or block production process in that case. WDYT?

Author: I added your sentence to the section

* **availability**: Yes (without SD, signer cannot register keys nor validly use other signers' keys)
* **integrity**: Yes (same, inaccurate SD will make key registration and signing process invalid)
* **confidentiality**: No

#### Mithril signing keys

SPOs register their Mithril keys every epoch to be able to sign snapshots. An attacker could impersonate the SPO and sign invalid snapshots if they got hold of those keys.
Signing keys are currently stored on-disk (?) 

* **confidentiality**: Yes 
* **integrity**: Yes (invalid key is useless obviously)
* **availability**: Yes (Signer needs Key at every signing round, unavailability will lead to inability to sign)

#### Mithril signing keys registration

Mithril signer needs to register new key every epoch with aggregator (and ultimately other signers)

* **confidentiality**: Yes (access to a signer's key will allow an attacker to impersonate a signer for the duration of the epoch)

> * **integrity**: Yes (partial? key registration process is transient and limited in time) 

Reviewer 2: If we are talking about all epochs, it may be called partial. But IMO, complete integrity is needed for a specific epoch. WDYT?

Author: Done

* **availability**: Yes (need access to aggregator to register key) 

#### Mithril signatures diffusion

Mithril signers produces signatures every time a new message needs to be signed (e.g. when a new immutable file is created in the Chain DB). Those signatures are generated from a random lottery based on the signer's stake and the protocol parameters.
Preventing Mithril signers from signing decreases the number of signatures and could allow attacker to take control of the produced snapshot

* **confidentiality**: No 
* **availability**: Yes
* **integrity**: No (signatures are considered tamper-proof)

#### Mithril protocol parameters 

Protocol parameters are needed to coordinate the production of valid multi-signatures. They are served by the aggregator

* **confidentiality**: No (they actually need to be public)

> * **availability**: Yes (partial?)

Author: Done

* **integrity**: Yes (tampering protocol parameters can lead a signer to produce invalid signatures)

> #### Era configuration files

Reviewer: This era-related assets should be captured by the system analysis section. Either link to a piece of documentation or explain there how this works and what it is used for.

Author: I added a link to the corresponding ADR :+1:

The era reader Era address is used by signers to extract information about the current Mithril Era which defines the structure of snapshots and therefore signatures.
It is stored in [GitHub](https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/release-mainnet/era.addr) and only modifiable through an approved merged PR

* **confidentiality**: No (they actually need to be public)
* **availability**: Yes
* **integrity**: Yes (tampering the files could lead to Mithril network not being able to create multi-signatures)

#### Era activation 

The current and next (if any) eras are announced on-chain with an era activation marker (see ADR https://mithril.network/doc/adr/4#era-activation-marker).

* **confidentiality**: No (they actually need to be public)
* **availability**: Yes (era markers are currently stored on-chain)
* **integrity**: Yes (tampering the era markers could lead to Mithril network not being able to create multi-signatures)

> #### Era verification vey

Reviewer 2: Typo: **key** :)

Author: Done

* **availability**: No 
* **confidentiality**: Yes
* **integrity**: Yes (?)

### Client-side only assets

These are the assets that are relevant only when downloading and verifying certificates and full-node snapshots. 

:::info To do

Verification process being trustless, does it really make sense to include those in the Threat Model?

:::

#### Mithril certificate verification process

Mithril clients download snapshots and verify associated certificates using the mithril-client library, either from a CLI tool or [embedded in a browser](https://mithril.network/explorer/) 

* **availability**: Yes (?)
* **confidentiality**: No
* **integrity**: No (the snapshots and certificates are assumed to be secure, integrity is inherent to the protocol)

#### Mithril snapshots & certificates

Mithril certificates are produced by the aggregator from the individual signatures. Mithril certificates' security rests upon a chain of trust ultimately pointing at the Genesis certificate. Mithril aggregator maintains snapshots and certificates (aggregated multi-signatures) to be served to clients

> * **availability**: No (Snapshots and certificates are just a fallback, a cardano-node client can always retrieve the infromation from the cardano network itself, albeit much more slowly)

Reviewer 4: type: client can always retrieve the inf**or**mation from the cardano network itself

Author: Done

* **confidentiality**: No
* **integrity**: No (the snapshots and certificates are assumed to be secure, integrity is inherent to the protocol)

#### Mithril Genesis Verification Key

Mithril genesis verification key is stored in [GitHub](https://github.com/input-output-hk/mithril/blob/main/mithril-infra/configuration/release-mainnet/genesis.vkey) and only modifiable through an approved merged PR

* **availability**: Yes (it's needed to verify a whole chain of certificate)
* **confidentiality**: No
* **integrity**: No (integrity is inherent to the protocol)

> #### Mithril Genesis Signing Key

Reviewer: Is this a "client-side only asset" (section this is in)? It sounds quite essential to the whole signing process?

Author: This is not a client-only asset: I moved it upper to the correct section :+1:

The corresponding signing key is stored in IOG's VaultWarden, and used only once, when the genesis certificate is generated. 

* **availability**: No (? The key is not needed unless a re-genesis process is required, but then a new key could be used instead?)
* **confidentiality**: Yes

> * **integrity**: Yes (?)

Reviewer 2: Do we say Genesis signing keys are tamper-proof?

Author: @curiecrypt could you make a suggestion for this comment?
(I think that it will also apply to the "Era signing key" section)

> ## Threat & Mitigations

Reviewer 3: Maybe one more: Currently the aggregation of single signatures from signers and the delivery of data to clients (certificate & artifacts) are both done by the aggregator.
An adversary could try to (D)DoS an aggregator signature aggregation process by sending requests from client(s), leading at worst to an epoch gap on the aggregator and the need to re-genesis the chain.

Author: I added a section for "Resource exhaustion on Mithril aggregator" :+1:

> ## Threat & Mitigations

Reviewer: Can you please add that this list is not exhaustive (just to over-communicate a bit)

Author: I added an info block stating the list is non exhaustive :+1:

### Resource exhaustion on Cardano relay

- DoS of a `mithril-signer` running next to a `cardano-node` being a relay
- Assets at risk: 
   - [Block diffusion](#block-diffusion)
   - [Mithril signatures diffusion](#mithril-signatures-diffusion)

### Resource exhaustion on Cardano block producer

* Assets at risk:
  - [Block production](#block-production)

### Block diffusion
* Diffusion is ensured through the connection between BPs, local relays, and downstream/upstream relays
* preventing them to operate can harm the Cardano network 
* Relay hosts connect the BP to the network, 
* Starving relay hosts prevents Mithril signatures and key registration to be shared
* Starving a cardano-node running on a relay host would prevent or delay the diffusion of new blocks thus harming 

> * Compromising relay hosts would be an extreme form of starving resources

Reviewer: Is this threat not the same as "Resource exhaustion on Cardano relay"?

Author: Yes, I moved this section as a sub-section of the "Resource exhaustion on Cardano relay" section :+1:

### Block production
* An incorrect mithril-signer could _starve_ the cardano-node of computing resources thus preventing it from producing and diffusing blocks in a timely manner

> * Compromising BP host would harms a BP's SPO economic viability

Reviewer: Is this threat not the same as "Resource exhaustion on Cardano block producer"?

### Hardening Operating System

> [Developers portal](https://developers.cardano.org/docs/operate-a-stake-pool/hardening-server) already provides thorough documentation on how to harden a linux-based host to run cardano-node

Reviewer: This clearly is a mitigation. Against which threat is this helping?

We are likely missing a threat and maybe even some assets, entry points, things to mention in the system analysis to motivate this mitigation (= its great to close some gaps by following this question)

Author: Indeed, the threat should be the "SPO's infrastructure security" for which a mitigation is "Hardening Operating System". WDYT?

Author: I moved it to the 'info' block of the section

---
sidebar_position: 3
sidebar_label: Mithril Threat Model
---

# Mithril Threat Model

:::danger

This document is a draft version of the **Mithril threat model** prepared by the **Mithril core team**:
- We expect to receive **external feedback and contributions** before we can consider it **final**.
- Feel free to **contribute** to this document by using the **Edit this page** link a the bottom of the page.

> - A **GitHub discussion** is also available [here](https://github.com/input-output-hk/mithril/discussions/).

Author: ```suggestion
- If you think there is a security vulnerability in Mithril, please disclose it responsibly by following the [Security Vulnerability Disclosure Policy](https://github.com/input-output-hk/mithril/blob/main/SECURITY.md)
```

Author: Done

* **availability**: No (Certificates are just a fallback, a Cardano node client can always retrieve the information from the Cardano network itself, albeit much more slowly)

#### Mithril artifacts

Mithril artifacts are produced by the aggregator specifically for the type of data being signed (e.g. a compressed snapshot archive of the Cardano node database). Mithril aggregator maintains artifacts to be served to clients.

* **confidentiality**: No
* **integrity**: No (the snapshots are verified against their associated certificate to establish their authenticity)
* **availability**: No (Artifacts are just a fallback, a Cardano node client can always retrieve the information from the Cardano network itself, albeit much more slowly)

#### Mithril genesis verification key

Mithril genesis verification key is stored in [GitHub](https://github.com/input-output-hk/mithril/blob/main/mithril-infra/configuration/release-mainnet/genesis.vkey) and only modifiable through an approved merged PR

* **confidentiality**: No
* **integrity**: No (integrity is inherent to the protocol)
* **availability**: Yes (it's needed to verify a whole chain of certificate)

## Threat & Mitigations

:::info

> This list of threat and mitigations is not exhaustive.

Author: ```suggestion
- This list of threat and mitigations is not exhaustive.
- [Developers portal](https://developers.cardano.org/docs/operate-a-stake-pool/hardening-server) already provides thorough documentation on how to harden a linux-based host to run cardano-node
```

Author: Done
